ivy/functional/ivy/general.py

```diff
@@ -2901,239 +2887,32 @@ def set_item(
+    query_ = tuple([q.to_numpy() if ivy.is_array(q) else q for q in query])
 
-def _parse_ellipsis(so, ndims):
-    pre = list()
-    for s in so:
-        if s is Ellipsis:
-            break
-        pre.append(s)
-    post = list()
-    for s in reversed(so):
-        if s is Ellipsis:
-            break
-        post.append(s)
-    ret = list(
-        pre
-        + [slice(None, None, None) for _ in range(ndims - len(pre) - len(post))]
-        + list(reversed(post))
-    )
-    return ret, (len(pre), ndims - len(post))
+    # array containing all of x's flat indices
+    x_ = ivy.arange(0, _numel(x_shape)).reshape(x_shape)
 
+    # use numpy's __getitem__ to get the queried indices
+    x_idxs = ivy.array(x_.to_numpy()[query_])
+    target_shape = x_idxs.shape
 
-def _parse_slice(idx, s):
-    step = 1 if idx.step is None else idx.step
-    if step > 0:
-        start = 0 if idx.start is None else idx.start
-        if start >= s:
-            stop = start
-        else:
-            if start <= -s:
-                start = 0
-            elif start < 0:
-                start = start + s
-            stop = s if idx.stop is None else idx.stop
-            if stop > s:
-                stop = s
-            elif start <= -s:
-                stop = 0
-            elif stop < 0:
-                stop = stop + s
-    else:
-        start = s - 1 if idx.start is None else idx.start
-        if start < -s:
-            stop = start
-        else:
-            if start >= s:
-                start = s - 1
-            elif start < 0:
-                start = start + s
-            if idx.stop is None:
-                stop = -1
-            else:
-                stop = idx.stop
-                if stop > s:
-                    stop = s
-                elif stop < -s:
-                    stop = -1
-                elif stop == -s:
-                    stop = 0
-                elif stop < 0:
-                    stop = stop + s
-    q_i = ivy.arange(start, stop, step).to_list()
-    q_i = [q for q in q_i if 0 <= q < s]
-    q_i = (
-        ivy.array(q_i)
-        if len(q_i) or start == stop or idx.stop is not None
-        else ivy.arange(0, s, 1)
-    )
-    return q_i
+    if 0 in x_idxs.shape or 0 in x_shape:
+        return None, target_shape
 
+    # convert the flat indices to multi-D indices
+    x_idxs = ivy.unravel_index(x_idxs, x_shape)
 
-def _deep_flatten(iterable):
-    def _flatten_gen(iterable):
-        for item in iterable:
-            if isinstance(item, list):
-                yield from _flatten_gen(item)
-            else:
-                yield item
+    # stack the multi-D indices to bring them to gather_nd/scatter_nd format
+    x_idxs = ivy.stack(x_idxs, axis=-1).astype(ivy.int64)
 
-    return list(_flatten_gen(iterable))
+    return x_idxs, target_shape
 
 
 def _numel(shape):
+    shape = tuple(shape)
     return math.prod(shape) if shape != () else 1
```

Comment:
```suggestion
    return ivy.prod(shape) if shape != () else 1
```
This might be better off being `ivy.prod` rather than `math.prod`? Not totally sure though, just a thought